### PR TITLE
Rework ResourcePool

### PR DESCRIPTION
**Pull request by Bolodya1997**
Config:
\- hostname
\+ services supported by PF

ResourcePool:
\- select VF by PF, IOMMU group
\- get host info
\+ select VF by driver type, capability, service

\- HostInfo
\- hostInfoServer

### CODE REVIEW

File: pkg/sriov/resourcepool/config.go [hunk 1]
```diff
@@ -27,10 +28,23 @@ import (
...
-// Config contains list of available physical functions and their capabilities
+// Config contains list of available physical functions
 type Config struct {
-	HostName          string                      `yaml:"hostName"`
-	PhysicalFunctions map[string]sriov.Capability `yaml:"physicalFunctions"`
+	PhysicalFunctions map[string]*PhysicalFunction `yaml:"physicalFunctions"`
+}
+
+func (c *Config) String() string {
+	out := "&{PhysicalFunctions:map["
```
Comment (denis-tingajkin) [inline pkg/sriov/resourcepool/config.go:37]: Please use strings.Builder

File: pkg/sriov/resourcepool/resource_pool.go [hunk 2]
```diff
@@ -29,179 +28,145 @@ import (
...
 
-			if pf, ok := rp.physicalFunctions[vf.PhysicalFunctionPCIAddress]; !ok {
-				rp.physicalFunctions[vf.PhysicalFunctionPCIAddress] = &physicalFunction{
-					capability: capability,
-					virtualFunctions: map[uint][]string{
-						vf.IommuGroupID: {
-							vf.PCIAddress,
-						},
-					},
+			switch pf, ok := rp.physicalFunctions[vf.PhysicalFunctionPCIAddress]; {
```
Comment (denis-tingajkin) [inline pkg/sriov/resourcepool/resource_pool.go:55]: Do we really need to use `switch `in the case when we can use only one `if` statement?
```suggestion

			pf, ok := rp.physicalFunctions[vf.PhysicalFunctionPCIAddress];
			if !ok {
			...
			}
			pf.virtualFunctions[vf.IommuGroupID] = append(pf.virtualFunctions[vf.IommuGroupID], vf.PCIAddress)
			
```

**Review by denis-tingajkin — COMMENTED** [review]
In general looks good, added few comments

File: pkg/sriov/resourcepool/resource_pool.go (hunk 2 above)
Comment (Bolodya1997) [reply pkg/sriov/resourcepool/resource_pool.go]: Really don't know why have I made it in such way :D

File: pkg/sriov/resourcepool/config.go (hunk 1 above)
Comment (Bolodya1997) [reply pkg/sriov/resourcepool/config.go]: Done

File: pkg/sriov/resourcepool/resource_pool.go [hunk 3]
```diff
@@ -29,179 +28,146 @@ import (
...
-	for pciAddr, pf := range rp.physicalFunctions {
-		pfInfo := &sriov.PhysicalFunctionInfo{
-			Capability:  pf.capability,
-			IommuGroups: make(map[uint]*sriov.IommuGroupInfo, len(pf.virtualFunctions)),
+	sort.Slice(vfs, func(i, k int) bool {
+		iIg := rp.iommuGroups[vfs[i].IommuGroupID]
+		kIg := rp.iommuGroups[vfs[k].IommuGroupID]
+		iPf := rp.physicalFunctions[vfs[i].PhysicalFunctionPCIAddress]
+		kPf := rp.physicalFunctions[vfs[k].PhysicalFunctionPCIAddress]
+		cmp := iPf.capability.Compare(kPf.capability)
```
Comment (denis-tingajkin) [inline pkg/sriov/resourcepool/resource_pool.go:97]: It looks like in cases when `iIg == driverType && kIg == sriov.NoDriver` we do not need to do `Compare(kPf.capability)`. I would suggest using here `if` statements to reduce the count of capability comparisons.

Comment (Bolodya1997) [reply pkg/sriov/resourcepool/resource_pool.go]: OK, I have split this switch into 2 parts.

File: pkg/sriov/resourcepool/resource_pool.go [hunk 4]
```diff
@@ -29,179 +28,149 @@ import (
...
+		switch {
+		case cmp < 0:
+			return true
+		case cmp > 0:
+			return false
+		default:
+			return iPf.freeVirtualFunctionsCount > kPf.freeVirtualFunctionsCount
```
Comment (denis-tingajkin) [inline pkg/sriov/resourcepool/resource_pool.go:111]: Can we extract this to  `Compare` method for physicalFunctions?

File: pkg/sriov/resourcepool/resource_pool.go [hunk 5]
```diff
@@ -29,179 +28,149 @@ import (
...
-func (rp *ResourcePool) Free(vfPciAddr string) {
-	logEntry := log.Entry(rp.ctx).WithField("ResourcePool", "Free")
-
-	vf := rp.vfByPciAddr(vfPciAddr)
+func (rp *ResourcePool) Free(vfPciAddr string) error {
+	vf := rp.findByPciAddr(vfPciAddr)
 	if vf == (*VirtualFunction)(nil) {
-		logEntry.Warnf("VF doesn't exist: %v", vfPciAddr)
-		return
+		return errors.Errorf("VF doesn't exist: %v", vfPciAddr)
```
[the remark this answers is not in the document]
Comment (Bolodya1997) [reply pkg/sriov/resourcepool/resource_pool.go:149]: Error strings should not be capitalized **(unless beginning with proper nouns or acronyms)** - it just starts with **VF** :)

File: pkg/sriov/resourcepool/resource_pool.go (hunk 4 above)
Comment (Bolodya1997) [reply pkg/sriov/resourcepool/resource_pool.go]: Yes we can, looks much more better now.

**Pull request merged**